clippy_lints/src/transmute/mod.rs

```diff
@@ -520,6 +521,32 @@ declare_clippy_lint! {
     "eager evaluation of `transmute`"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks if transmute calls have all generics specified.
+    ///
+    /// ### Why is this bad?
+    /// If not set, some unexpected output type could be retrieved instead of the expected one,
+    /// potentially leading to invalid code.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// # unsafe {
+    /// let x: i32 = std::mem::transmute([1u16, 2u16]);
+    /// # }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// # unsafe {
+    /// let x = std::mem::transmute::<[u16; 2], i32>([1u16, 2u16]);
+    /// # }
+    /// ```
+    #[clippy::version = "1.77.0"]
+    pub MISSING_TRANSMUTE_ANNOTATIONS,
+    suspicious,
```

Comment:
I agree with your analysis. Even if the `transmute` as the last statement/tail expression, it doesn't hurt to have the types information right on the `transmute` without needing to go back up to the function declaration (for long functions). And if there is a change in the returned type, it'll force users to come back to this code to ensure it's still what they want.

Reply:
Ah, I just looked at the other PR, and saw that @xFrednet you had a similar concern here: https://github.com/rust-lang/rust-clippy/pull/11047#discussion_r1252221990 , do you still think it should be allow-by-default?

I'm of the opinion that this would generally be useful to have, as warn-by-default, since the advantage of protecting against type inference for such a heavy hammer (`transmute`) IMO outweigh the occasional situation-dependent `#[allow]` one might need. I'm interested in hearing others' opinions, tho.

Reply:
This feels like a lint that should be warn-by-default. Though, we should run a lintcheck to see how many reports it generates. If it's too many, then it might be better to have it in `pedantic`. If we generate a lot of new warnings, it might be counterproductive and cause users to just allow the lint without thinking about the consequences.

Reply:
I don't think "it generates too many warnings" is a good point to not keep it as warn-by-default, quite the opposite. This lint isn't emitted if it's generated from an external macro or if the result of the `transmute` is used directly in a let binding with type annotation. Anything else should most definitely warn as it could allow to uncover some problematic code.

Reply:
> I don't think "it generates too many warnings" is a good point to not keep it as warn-by-default

It can be hard to pick the correct lint group/level, it depends on the lint and what it detects. As said, just by reading this, I would consider it to be a warn-by-default lint. Running lintcheck is meant to confirm this feeling.

I'd still argue that the number of warnings should be a factor or at least be looked at for mainly two reasons:

1. If there are too many warnings, it's an indication that the suggested change might not be idiomatic.
2. We have to think about the users. If Clippy generates too many errors/warnings, users are more prune to just disabling a lint and being annoyed. There were a few lints, I would have considered being warn-by-default, but the community disagreed with. (Which is totally fine, discussions on these topics are important) I believe that some of these cases were not because of the lint, but because of the amount of required changes.

---

Generally speaking and slightly off-topic, I would hope that most projects enable the `pedantic` group as well. For small hacking scrips it might be overkill, but on bigger projects I found it to be a big help.

Reply:
\o/

Reply:
Out of curiosity: how many crates was it checking? I also ran it earlier and got 16 hits, but that was with 200 crates. (Not that it changes my view on the category; I still agree)

Reply:
26 as far as I can see.

Reply:
@y21 I think it might be worth it to update the default list in this repo to include a larger set like yours.

Reply:
Yeah, I agree. I think the default list is often not large enough to come to useful conclusions, there has been instances where the default list hasn't found anything but 200 found a bunch. But it also requires some manual reviewing if we do that I think (instead of blindly using lintcheck's `popular-crates` subcommand). For example, in one crate, there are *200,000* hits for the match_same_arm lint.